core/src/crds.rs

```diff
@@ -275,22 +280,52 @@ impl Crds {
         now: u64,
         timeouts: &HashMap<Pubkey, u64>,
     ) -> Vec<CrdsValueLabel> {
-        // TODO: need custom logic for purging duplicate shreds.
+        #[rustversion::before(1.49.0)]
+        fn select_nth<T: Ord>(records: &mut Vec<T>, _nth: usize) {
+            records.sort_unstable();
+        }
+        #[rustversion::since(1.49.0)]
+        fn select_nth<T: Ord>(records: &mut Vec<T>, nth: usize) {
+            records.select_nth_unstable(nth);
+        }
         let default_timeout = *timeouts
             .get(&Pubkey::default())
             .expect("must have default timeout");
-        thread_pool.install(|| {
-            self.table
-                .par_iter()
-                .with_min_len(1024)
-                .filter_map(|(k, v)| {
-                    let timeout = timeouts.get(&k.pubkey()).unwrap_or(&default_timeout);
-                    if v.local_timestamp.saturating_add(*timeout) <= now {
-                        Some(k.clone())
-                    } else {
+        let evict = |pubkey, index: &IndexSet<usize>| {
+            let timeout = *timeouts.get(pubkey).unwrap_or(&default_timeout);
+            let mut old_labels = Vec::new();
+            let mut buffer: Vec<_> = index
```

Comment:
Took me a while to read this. There might be another name for `buffer` which is more descriptive. And maybe some comments either here and/or the function signature about "Find old or extra labels" and "Prune extra labels from the same pubkey".

Reply:
Added more comments and type annotations.
The `buffer` is literally buffering indices to be partitioned. I added comment for clarity but if you have suggestion for a better name I will take it.